contracts/LimitOrderSwap.sol

```diff
@@ -70,6 +70,73 @@ contract LimitOrderSwap is ILimitOrderSwap, Ownable, TokenCollector, EIP712 {
         _fillLimitOrder(order, makerSignature, takerParams, true);
     }
 
+    /// @inheritdoc ILimitOrderSwap
+    function fillLimitOrderGroup(
+        LimitOrder[] calldata orders,
+        bytes[] calldata makerSignatures,
+        uint256[] calldata makerTokenAmounts,
+        address[] calldata profitTokens
+    ) external payable override {
+        if (orders.length != makerSignatures.length || orders.length != makerTokenAmounts.length) revert InvalidParams();
+
+        // validate orders and calculate takingAmounts
+        uint256 unwrapAmount;
+        uint256[] memory takerTokenAmounts = new uint256[](orders.length);
+        for (uint256 i = 0; i < orders.length; ++i) {
+            LimitOrder memory order = orders[i];
+            uint256 makerTokenAmount = makerTokenAmounts[i];
+
+            if (order.expiry <= block.timestamp) revert ExpiredOrder();
+            if (order.taker != address(0) && msg.sender == order.taker) revert InvalidTaker();
+
+            bytes32 orderHash = getLimitOrderHash(order);
+            {
+                uint256 orderFilledAmount = orderHashToMakerTokenFilledAmount[orderHash];
+                if (orderFilledAmount == 0) {
+                    // validate maker signature only once per order
+                    if (!SignatureValidator.isValidSignature(order.maker, getEIP712Hash(orderHash), makerSignatures[i])) revert InvalidSignature();
+                }
+
+                // check whether the order is fully filled or not
+                if (orderFilledAmount >= order.makerTokenAmount) revert FilledOrder();
+                // check whether the order is canceled or not
+                if ((orderFilledAmount & ORDER_CANCEL_AMOUNT_MASK) != 0) revert CanceledOrder();
+
+                uint256 orderAvailableAmount = order.makerTokenAmount - orderFilledAmount;
+                if (makerTokenAmount > orderAvailableAmount) revert NotEnoughForFill();
+                takerTokenAmounts[i] = ((makerTokenAmount * order.takerTokenAmount) / order.makerTokenAmount);
+                if (order.takerToken.isETH()) {
+                    unwrapAmount += takerTokenAmounts[i];
+                }
+
+                // record fill amount
+                orderHashToMakerTokenFilledAmount[orderHash] = orderFilledAmount + makerTokenAmount;
+            }
+
+            // collect maker tokens
+            _collect(order.makerToken, order.maker, address(this), makerTokenAmount, order.makerTokenPermit);
+
+            // transfer fee if present
+            uint256 fee = (makerTokenAmount * order.feeFactor) / Constant.BPS_MAX;
+            order.makerToken.transferTo(feeCollector, fee);
+
+            _emitLimitOrderFilled(order, orderHash, takerTokenAmounts[i], makerTokenAmount - fee, fee, address(this));
```

Comment:
It's hard to say that the `msg.sender` is the recipient. From a single order perspective, the recipient is the `LimitOrderSwap` order since it's where maker token transferred to.

Reply:
Because I see that the profitTokens are transferred to `msg.sender` so I assume `msg.sender` is the (profit) recipient. How do we extract the actual `recipient` off-chain in the group fill case?

Reply:
I mean the `recipient` of an order should be `LimitOrderSwap` contract and since the event is about an order gets filled, recipient should be `address(this)`. Considering profit, perhaps adding another event like `SendGroupFillProfit` and in that way the recipient should be `msg.sender`.